File: PracticasPDOORuby/Practica2_Ruby/lib/Irrgarten/player.rb
# frozen_string_literal: true
require_relative 'weapon'
require_relative 'shield'
require_relative 'dice'

module Irrgarten
  class Player
    # --- Constantes y Atributos de Lectura ---
    MAX_WEAPONS = 2
    MAX_SHIELDS = 3
    INITIAL_HEALTH = 10
    HITS2LOSE = 3

    attr_reader :row, :col, :name, :name, :intelligence, :strength, :health

    # ======================================================
    # MÉTODOS PÚBLICOS
    # ======================================================
    # Constructor
    # @param number Identificador del jugador (0, 1, 2, ...)
    # @param intelligence Nivel de inteligencia del jugador
    # @param strength Nivel de fuerza del jugador
    # @return Objeto de la clase Player
    def initialize(number, intelligence, strength)
      @number = number.to_s[0]
      @intelligence = intelligence.to_f
      @strength = strength.to_f
      @name = "Player#{@number}"
      @health = INITIAL_HEALTH.to_f
      @row = @col = -1
      @consecutive_hits = 0
      @weapons = []
      @shields = []
    end

    # Resucita al jugador, dejándolo como al principio
    # @return Nada (modifica el estado del jugador)
    def resurrect
      @weapons.clear
      @shields.clear
      @health = INITIAL_HEALTH
      @consecutive_hits = 0
    end

    # Establece la posición del jugador en el laberinto
    # @param row Fila de la posición
    # @param col Columna de la posición
    # @return Nada (modifica el estado del jugador)
    def set_pos(row, col) = (@row, @col = row.to_i, col.to_i)

    # Indica si el jugador está muerto
    # @return true si el jugador está muerto (salud <= 0), false en caso contrario
    def dead? = @health <= 0

    # Calcula el ataque del jugador (fuerza + armas)
    # @return Valor del ataque del jugador
    def attack
      @strength + sum_weapons
    end

    # Defiende al jugador de un ataque recibido
    # @param received_attack Valor del ataque recibido
    # @return Nada (modifica el estado del jugador)
    def defend(received_attack)
      manage_hit(received_attack)
    end

    # El jugador ha sido herido (pierde 1 punto de salud)
    # @return Nada (modifica el estado del jugador)
    def got_wounded
      @health -= 1
    end

    # Representación en cadena del jugador
    # @return Cadena con la información del jugador
    def to_s
      "Player{nombre = '#{@name}', inteligencia = '#{@intelligence}', fuerza = '#{@strength}', salud = '#{@health}', posicion = (#{@row}, #{@col}), armas = '#{@weapons}', escudos = '#{@shields}'"
    end

    # ======================================================
    # P3 - Métodos públicos a implementar
    # ======================================================
    def move(direction, valid_moves) = (raise NotImplementedError, "P3")
    def receive_reward = (raise NotImplementedError, "P3")

    # =====================================================
    # MÉTODOS PRIVADOS
    # =====================================================
    private

    # Calcula la energía defensiva del jugador (inteligencia + escudos)
    # @return Valor de la energía defensiva del jugador
    def defensive_energy
      @intelligence + sum_shields
    end

    # Resetea el contador de golpes consecutivos
    # @return Nada (modifica el estado del jugador)
    def reset_hits
      @consecutive_hits = 0
    end

    # Incrementa el contador de golpes consecutivos
    # @return Nada (modifica el estado del jugador)
    def inc_consecutive_hits
      @consecutive_hits += 1
    end

    # Suma el poder de ataque de todas las armas del jugador
    # @return Suma del poder de ataque de las armas
    def sum_weapons
      @weapons.sum {|w| w.attack}
    end

    # Suma la protección de todos los escudos del jugador
    # @return Suma de la protección de los escudos
    def sum_shields
      @shields.sum {|s| s.protect}
    end

    # Crea un nuevo arma con poder y usos aleatorios
    # @return Nuevo objeto de la clase Weapon
    def new_weapon
      Weapon.new(Dice.weapon_power, Dice.uses_left)
    end

    # Crea un nuevo escudo con poder y usos aleatorios
    # @return Nuevo objeto de la clase Shield
    def new_shield
      Shield.new(Dice.shields_power, Dice.uses_left)
    end

    # ======================================================
    # P3 - Métodos privados a implementar
    # ======================================================
    def receive_weapon(w) = (raise NotImplementedError, "P3")
    def receive_shield(s) = (raise NotImplementedError, "P3")
    def manage_hit(received_attack) = (raise NotImplementedError, "P3")
  end
end

File: PracticasPDOOJava/Practica3_Java/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;

/**
 * Representa un jugador dentro del laberinto
 * <p>
 * Cada jugador posee características propias como inteligencia, fuerza y salud,
 * además de una posición (row/column) dentro del {@link labyrinth}.
 * Puede equipar un número de limitado de armas y escudos, atacar, defenderse y 
 * resucitar tras ser derrotado.
 * </p>
 * 
 * @author Joaquín Cruz Lorenzo
 */
public class Player {
    // ==================================
    // CONSTANTES DE CLASE
    // ==================================
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    // ==================================
    // ATRIBUTOS DE INSTANCIA 
    // ==================================
    private String name;
    private char number;
    private float intelligence, strength, health;
    private int row, col, consecutiveHits = 0;
    
    private ArrayList<Weapon> weapons = new ArrayList();
    private ArrayList<Shield> shields = new ArrayList();
    
    // ===============================================
    // MÉTODOS PÚBLICOS
    // ===============================================
    
    /**
     * Constructor
     * @param number
     * @param intelligence
     * @param strength 
     */
    public Player(char number, float intelligence, float strength) {
        this.number = number;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = INITIAL_HEALTH;
        this.row = -1;
        this.col = -1;
        this.name = "Player #" + number;
    }
    
    /**
     * Restaura al jugador al estado inicial.
     * <ul>
     *          <li> Vacía armas y escudos. </li>
     *          <li> Reinicia salud e impactos consecutivos. </li>
     * </ul>
     */
    public void resurrect() {
        weapons.clear();
        shields.clear();
        health = INITIAL_HEALTH;
        consecutiveHits = 0;
    }

    /**
     * 
     * @return fila actual del jugador 
     */
    public int getRow() {
        return row;
    }
    
    /**
     * 
     * @return columna actual del jugador
     */
    public int getCol() {
        return col;
    }
    
    /**
     * 
     * @return número del jugador 
     */
    public char getNumber() {
        return number;
    }
    
    /**
     * Cambia la posicion del jugador dentro del laberinto.
     * @param row
     * @param col 
     */
    public void setPos(int row, int col) {
        this.row = row;
        this.col = col;
    }
    
    /**
     * Indica si el jugador está muerto (salud menor o igual a 0).
     * @return true si el jugador está muerto, false en caso contrario.
     */
    public boolean dead() {
       return health <= 0;
    }
    
    /**
     * Calucla la potencia de ataque del jugador.
     * <p> La fórmula básica es {@code strength + sumWeapons()} </p>
     * @return  potencia de ataque total
     */
    public float attack() {
        return strength + sumWeapons();
    }
    
    /**
     * Inicia la defensa ante un ataque recibido.
     * <p> Lo dejamos sin implementar en P2. Se nos dará información en P3 </p>
     * @param receivedAttack intensidad del ataque recibido
     * @return 
     */
    public boolean defend(float receivedAttack) {
        throw new UnsupportedOperationException();
    }
    
    @Override
    public String toString() {
        return "Player{" + "name= " + name + ", number= " + number + ", intelligence=" + intelligence + ", strength= " + strength + ", health= " + health + 
                ", row= " + row + ", col= " + col + ", consecutiveHits= " + consecutiveHits + ", weapons= " + weapons + ", shields= " + shields + '}';
    }
    
    // ==================================
    // P3 - Métodos públicos a implementar
    // ==================================
    public Directions move(Directions direction, Directions[] validMoves){ throw new UnsupportedOperationException(); }
    public void receiveReward() { throw new UnsupportedOperationException(); }
    
    
    
    // ==========================================================
    // MÉTODOS PRIVADOS
    // ==========================================================
    /**
     * Suma la potencia de todas las armas equipadas (consumiendo sus usos)
     * @return suma total de ataque de armas activas
     */
    private float sumWeapons() {
        float total = 0.0f;
        
        for (Weapon w : weapons) total += w.attack();
        
        return total;
    }
    
    /**
     * Suma la capacidad de protección de todos los escudos (consumiendo sus usos)
     * @return  protección total ofrecida por los escudos
     */
    private float sumShields() {
        float total = 0.0f;
        
        for (Shield s : shields) total += s.protect();
        
        return total;
    }
    
    /**
     * Calcula la energía defensiva del jugador.
     * <p> {@code defensiveEnergy = intelligence + sumShields()} </p>
     * @return  valor de energía defensiva
     */
    private float defensiveEnergy() {
        return intelligence + sumShields();
    }
    
    /**
     * Incrementa el número de impactos consecutivos sufridos.
     */
    private void incConsecutiveHits() {
        consecutiveHits++;
    }
    
    /**
     * Reinicia el contador de impactos consecutivos a 0.
     */
    private void resetHits() {
        consecutiveHits = 0;
    }
    
    /**
     * Resta un punto de salud al jugador.
     */
    private void gotWounded() {
        health--;
    }
    
    /**
     * Genera un arma nueva con valores aleatorios dados por {@link Dice}.
     * @return  nueva instancia de {@link Weapon}
     */
    private Weapon newWeapon() {
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    /**
     * Genera un escudo nuevo con valores aleatorios dados por {@link Dice}.
     * @return nueva instancia de {@link Shield}
     */
    private Shield newShield() {
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }
    
    // =========================================
    // P3 - Métodos privados a implementar
    // =========================================
    private void receiveWeapon(Weapon w) { throw new UnsupportedOperationException(); }
    private void receiveShield(Shield s) { throw new UnsupportedOperationException(); }
    private boolean manageHit(float receivedAttack) { throw new UnsupportedOperationException(); }
}
